src/daemons/linuxd/src/linuxd/mod.rs

```diff
@@ -762,50 +816,55 @@ impl LinuxDaemon {
-                if request.count == 0 {
-                    // Writing zero-bytes to STDOUT is not allowed, as we used this to signal EOF.
-                    error!("handle_write_request(): trying to write zero bytes to STDOUT");
-                    build_error(source, ErrorCode::InvalidArgument)
-                } else {
-                    profiler::timestamp_message!(&mut request.buffer, 0);
-                    // NOTE: we don't check if the write operation is too big, because its size is
-                    // already bound by the maximum payload size of the message.
-                    let count: usize = request.count as usize;
-                    match conn.send_message_to_gateway(&request.buffer[..count]) {
-                        Ok(_) => {
-                            debug!("wrote {count} bytes to the gateway");
-                            WriteResponse::build(source, count as i32)
-                        },
-                        Err(e) => {
-                            debug!("failed to write buffer to the gateway (error={e:?})");
-                            // TODO: Check error conversion.
-                            build_error(source, ErrorCode::ConnectionReset)
-                        },
-                    }
-                }
+            let gateway_stdout_tx = if let Some(gateway_stdout_tx) = gateway_stdout_tx {
+                gateway_stdout_tx
+            } else {
+                error!("handle_write_request(): trying to write to stdout without a gateway configured");
+                return build_error(source, ErrorCode::InvalidArgument);
+            };
+
+            let mut venv = venv.lock().unwrap();
+            let env = if let Some(env) = venv.get_mut(source) {
+                env
             } else {
-                // Not connected to the gateway, print to stdout.
+                error!("handle_write_request(): thread not assigned to venv");
+                return build_error(source, ErrorCode::InvalidArgument);
+            };
+
+            // Check if write size is invalid.
+            if request.count == 0 {
+                // Writing zero-bytes to STDOUT is not allowed, as we used this to signal EOF.
+                error!("handle_write_request(): trying to write zero bytes to STDOUT");
+                build_error(source, ErrorCode::InvalidArgument)
+            } else {
+                profiler::timestamp_message!(&mut request.buffer, 0);
                 let count: usize = request.count as usize;
-                let buffer: &[u8] = &request.buffer[..count];
-                if request.fd == STDERR_FILENO {
-                    let _ = io::stderr().write_all(buffer);
-                    let _ = io::stderr().lock().flush();
-                } else {
-                    let _ = io::stdout().write_all(buffer);
-                    let _ = io::stdout().lock().flush();
+                if let Err(error) = gateway_stdout_tx.send((request, env.get_stdout_response_tx())) {
```

Comment:
@ppenna here we are being slightly conservative with writes to stdout.

We first send the write request (`request`) then wait for the IO thread to push an (empty) response confirming it has written the value to the output stream (`env.get_stdout_response_rx().recv()` in the `match` statement below).

Alternatively, we could not wait at all for the IO thread to send a response, this would also save us having to pass the `venv` to `handle_write_request`. Given that the write operation on the gateway socket is also asynchronous, I think it is safe to not wait for the IO thread's response, but I wanted to know your take before implementing it.